one/src/cbor_value.rs

```diff
@@ -0,0 +1,285 @@
+                let tag: Tag = decoder.tag()?;
+                let value = CborValue::next(decoder)?;
+                Ok(CborValue::Tag((tag, Box::new(value))))
+            }
+            Ok(Type::Break) => Ok(CborValue::Break),
+            Ok(Type::Unknown(additional)) => Ok(CborValue::Unknown(additional)),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    /// Accesses a deeply nested CborValue when the path keys are all strings.
+    pub fn path(&self, parts: &[&str]) -> CborValue {
+        match parts.split_first() {
+            None => self.clone(), // if there are no parts this is thing at the path.
+            Some((first, rest)) => {
+                let CborValue::Map(map) = &self else {
+                    return CborValue::Undefined; // if self is not a map there is no thing at the path.
+                };
+                let Some(next) = map.get(&first.to_string().into()) else {
+                    return CborValue::Undefined; // if the next part is not in the map there is no thing at the path.
+                };
+                next.path(rest)
+            }
+        }
+    }
+
+    /// Accesses the underlying CborValue for a specified tag type.
+    pub fn untag(&self, tag: u64) -> Option<&CborValue> {
+        if let CborValue::Tag((Tag::Unassigned(inner_tag), boxed_value)) = &self {
+            if inner_tag == &tag {
+                return Some(boxed_value);
+            }
+        }
+        None
+    }
+
+    /// get a printable string of the type_name
+    pub fn type_name(&self) -> String {
+        match self {
+            CborValue::Bool(_) => "bool".to_string(),
+            CborValue::Null => "null".to_string(),
+            CborValue::Undefined => "undefined".to_string(),
+            CborValue::U64(_) => "u64".to_string(),
+            CborValue::I64(_) => "i64".to_string(),
+            CborValue::Int(_) => "int".to_string(),
+            CborValue::F64(_) => "f64".to_string(),
+            CborValue::Simple(_) => "simple".to_string(),
+            CborValue::Bytes(_) => "bytes".to_string(),
+            CborValue::String(_) => "string".to_string(),
+            CborValue::Array(_) => "array".to_string(),
+            CborValue::Map(_) => "map".to_string(),
+            CborValue::Tag(_) => "tag".to_string(),
+            CborValue::Break => "break".to_string(),
+            CborValue::Unknown(_) => "unknown".to_string(),
+        }
+    }
+
+    /// If the CborValue is a CborValue::Array get the CborValue at index.
+    /// If it is not a CborValue::Array or index is out of bounds returns None.
+    pub fn get_index(&self, index: usize) -> Option<&CborValue> {
+        self.as_array()?.get(index)
+    }
+
+    /// If the CborValue is a CborValue::Map get the CborValue at index.
+    /// If it is not a CborValue::Map or key is missing returns None.
+    pub fn get_key(&self, key: &str) -> Option<&CborValue> {
+        self.as_map()?.get(&CborValue::String(key.to_owned()))
```

Comment:
my general API sense is that you should try to avoid allocating internally as the "principle of least surprise". since you need an owned string, my opinion is we should take `key: String` but it's fine.

also, you could potentially use a `Cow<'_, str>` as the variant body

Reply:
https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.get

```rs
pub fn get<Q>(&self, key: &Q) -> Option<&V>
where
    K: Borrow<Q> + Ord,
    Q: Ord + ?Sized,
```

If we can make an `==` for `&str` and `CborValue::String` we may be able to avoid making a `CborValue::String` to use `std::collections::BTreeMap.get(key)`.

Reply:
Yeah, we could do that with Deref but it's probably not worth it (will possibly have some surprising behavior). 

It's fine to close this, I was just commenting on the api signature generally. My opinion is to use `Into<String>` or `String` rather than `&str` and allocating internally. 

If we create and share/copy a lot of `String(String)` variants, changing it to `String(Cow<_,str>)` might help and not require a struct lifetime but I'm not sure how easy it'd be to work with. My point was just that you could create that from the `&str` without needing `to_owned()`.